inc/az_ulib_registry_api.h

```diff
@@ -0,0 +1,198 @@
+
+/**
+ * @file
+ *
+ * @brief   The Registry Editor.
+ *
+ * The Registry Editor is capable of storing key value pairs into device flash.
+ */
+
+#ifndef AZ_ULIB_REGISTRY_API_H
+#define AZ_ULIB_REGISTRY_API_H
+
+#include "az_ulib_pal_api.h"
+#include "az_ulib_result.h"
+#include "azure/az_core.h"
+
+#ifdef __cplusplus
+#include <cstddef>
+#include <cstdint>
+#else
+#include <stddef.h>
+#include <stdint.h>
+#endif
+
+#include "azure/core/_az_cfg_prefix.h"
+
+/**
+ * @brief   Registry control block.
+ *
+ *  Internal structure to control the registry.
+ */
+typedef struct
+{
+  /** Pointer to the start of the memory to store the registry. */
+  void* registry_start;
+
+  /** Pointer to the end of the memory to store the registry. */
+  void* registry_end;
+
+  /** Pointer to the start of the memory to store the registry information. */
+  void* registry_info_start;
+
+  /** Pointer to the end of the memory to store the registry information. */
+  void* registry_info_end;
+
+  /** Size of each page. */
+  size_t page_size;
+
+} az_ulib_registry_control_block;
+
+/**
+ * @brief   Registry information.
+ *
+ *  Structure to return registry information in the az_ulib_registry_get_info().
+ */
+typedef struct
+{
+  /** Total number of registry info that can be stored in the Registry. In number of registries. */
+  size_t total_registry_info;
+
+  /** Current number of registries stored in the Registry. In number of registries. */
+  size_t in_use_registry_info;
+
+  /** Number of free registry info positions. In number of registries. */
+  size_t free_registry_info;
+
+  /** Total memory reserved to store registry data in bytes. */
+  size_t total_registry_data;
+
+  /** Total memory currently used to store registry data in bytes. */
+  size_t in_use_registry_data;
+
+  /** Total free memory to store registry data In bytes. */
+  size_t free_registry_data;
+} az_ulib_registry_info;
+
+/**
+ * @brief   This function gets the #az_span value associated with the given #az_span key from the
+ * registry.
+ *
+ * This function goes through the registry comparing keys until it finds one that matches the input.
+ *
+ * @param[in]   key                 The #az_span key to look for within the registry.
+ * @param[out]  value               The point to #az_span value corresponding to the input key.
+ *
+ * @pre         Registry shall already be initialized.
+ * @pre         \p key              shall not be `#AZ_SPAN_EMPTY`.
+ * @pre         \p value            shall not be `NULL`.
+ *
+ * @return The #az_result with the result of the registry operations.
+ *      @retval #AZ_OK                        If retrieving a value from the registry was
+ *                                            successful.
+ *      @retval #AZ_ERROR_ITEM_NOT_FOUND      If there are no values that correspond to the
+ *                                            given key within the registry.
+ */
+AZ_NODISCARD az_result az_ulib_registry_try_get_value(az_span key, az_span* value);
```

Comment:
I feel like "try" in this case implies that this function could return AZ_ERROR_ULIB_BUSY. Also I think the fact that you are getting a value is implied.

Consider:
```c
AZ_NODISCARD az_result az_ulib_registry_get(az_span key, az_span* value);
```

Reply:
this naming is also in line with the following api `az_ulib_registry_add()`

Reply:
"try" is not because it could return BUSY, but because luck up the provided `key` can fail, for example if it is not found.

"value" is because we can decided to add another get in the future, so we can do that without have an ambiguous name or a breaking change.

Reply:
#Wontfix

Reply:
Any strong opinion about it?

Reply:
I feel that it is standard for get functions on data structures like maps, tables, and trees to accept a key as an input and return a value as an output. As for the "try: part, I'm still a bit confused because the way you have defined it, any "get" function should be a "try_get." I think it is implied that a "get" function can fail if the provided key is not found.

Reply:
After chat offline, we decided to keep the original name. #Wontfix